### PR TITLE
Proposal: add new `headers` event

### LINKED ISSUE #171: Accessing request.form in a @app.listen("request") raises a LazyBody error.
I don't really know if it is a bug or a feature.

I was logging the request parameters using a `@app.listen(request)` decorator but at this point the request.body is not loaded yet.
I changed it for `@app.listen("response")`

### PR DESCRIPTION
**Pull request by yohanboniface**
fix #171

### CODE REVIEW

File: tests/test_hooks.py [hunk 1]
```diff
@@ -83,6 +83,54 @@ async def listener(request, response):
...
+    assert resp.status == HTTPStatus.OK
+    assert resp.body == b'raised as expected'
+
+
+async def test_headers_hook_can_consume_request_body_explicitly(client, app):
+
+    @app.listen('headers')
+    async def listener(request, response):
+        response.status = 200
+        await request.load_body()
```
Comment (yohanboniface) [inline tests/test_hooks.py:108]: I find a bit ugly that we need two lines to consume a lazy body. My first approach was to write: `response.body = await request.read()`, but 1. there is no `read` method, 2 the `load_body` method does not return the body.

This has been changed [here](https://github.com/pyrates/roll/pull/89/commits/02c40531253d7c2b7676ba422acb489b475b5a75) but I see [here](https://github.com/pyrates/roll/pull/89#discussion_r357960321) that *I* pointed to this change (this shows how inconsistent I am :p )

I'm sort of tempted to revert that commit and have back a `read` method, which I find more natural, and I think it's explicit that a method that return something will but this something in memory.

Thoughts ? :)

File: docs/changelog.md [hunk 2]
```diff
@@ -8,6 +8,14 @@ A changelog:
 
 *Analogy: git blame :p*
 
+## dev
+
+- **Breaking change**:
+  - `request` event is now sent *after* the URL matching and the HTTP verb check,
+  please use new `headers` event for those use cases
```
Comment (davidbgk) [inline docs/changelog.md:15]:
```suggestion
  please use the new `headers` event for these use cases.
```

File: roll/__init__.py [hunk 3]
```diff
@@ -65,16 +65,17 @@ async def shutdown(self):
     async def __call__(self, request: Request, response: Response):
         payload = request.route.payload
         try:
-            if not await self.hook("request", request, response):
+            if not await self.hook("headers", request, response):
                 if not payload:
                     raise HttpError(HTTPStatus.NOT_FOUND, request.path)
-                # Uppercased in order to only consider HTTP verbs.
+                    # Uppercased in order to only consider HTTP verbs.
```
Comment (davidbgk) [inline roll/__init__.py:71]: Weird reindentation?

File: tests/test_hooks.py (hunk 1 above)
Comment (davidbgk) [reply tests/test_hooks.py:108]: Can you rewrite `it's explicit that a method that return something will but this something in memory` in English please :p

Comment (yohanboniface) [reply tests/test_hooks.py:108]: s/but/put/ ! :p

Comment (davidbgk) [reply tests/test_hooks.py:108]: If I remember correctly, we wanted to have an explicit read/load step for future usage/flexibility, but this is an old discussion and I'm not sure it's still useful… now that it's released do we want to introduce a breaking change for that? 🤔

Comment (yohanboniface) [reply tests/test_hooks.py:108]: No need for a breaking change:

```
async def read(self):
    await self.load_body()
    return self._body
```

or

```
async def read(self):
    …

load_body = read
```

Comment (davidbgk) [reply tests/test_hooks.py:108]: Indeed, I prefer the first approach, maybe we'll change our mind again at some point in the future 🙈

**Review by davidbgk — APPROVED** [review]

**Pull request merged**